Js_of_ocaml TodoMVC Example.

slegrand45 opened the pull request:
[Js_of_ocaml](http://ocsigen.org/js_of_ocaml/) is a compiler of OCaml bytecode to Javascript.

[OCaml](https://ocaml.org/) is a general purpose programming language with an emphasis on expressiveness and safety. Js_of_ocaml brings nearly all the OCaml features to Javascript. And especially its powerful type system.

arthurvr commented:
Personally I think should totally go discuss this at [TasteLang](https://github.com/tastejs/TasteLang) but that it isn't really a good fit for TodoMVC. Mainly because this is more about the language than about an MV* framework. What does the rest think?

passy commented:
OCaml has a really nice JS story these days and I hear it more and more, especially among people looking for a type-safe way to write web apps, where there aren't very many options out there. The implementation does a pretty nice job of separating the concerns and looks pretty similar to the Elm way, so I'm :+1:.

arthurvr commented on examples/js_of_ocaml/build.sh:
```diff
@@ -0,0 +1,16 @@
+
+js_of_ocaml +weak.js --opt 3 -o js/todomvc.js todomvc.byte
+
+# Avoid JS linters (hard to apply for auto generated JS)
```
Instead of this magic just ignore specific files in the config files.

arthurvr commented on examples/js_of_ocaml/index.html:
```diff
@@ -0,0 +1,13 @@
+    <div id="todomvc"></div>
+    <script src="js/todomvc.js"></script>
+  </body>
+</html>
```
missing newline

arthurvr commented on examples/js_of_ocaml/index.html:
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
+<html lang="en">
+  <head>
```
we use tabs for indentation.

arthurvr commented on examples/js_of_ocaml/index.html:
```diff
@@ -0,0 +1,13 @@
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Js_of_ocaml • TodoMVC</title>
+    <link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
```
This misses todomvc-common (check the other apps)

arthurvr commented on examples/js_of_ocaml/index.html:
```diff
@@ -0,0 +1,13 @@
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Js_of_ocaml • TodoMVC</title>
```
Let's properly write this instead of snake_case.

passy replied:
I think this is the official spelling.

On Fri, Jun 26, 2015, 22:48 Arthur Verschaeve <notifications@github.com>
wrote:

> In examples/js_of_ocaml/index.html
> <https://github.com/tastejs/todomvc/pull/1339#discussion_r33399033>:
>
> > @@ -0,0 +1,13 @@
> > +<!doctype html>
> > +<html lang="en">
> > +  <head>
> > +    <meta charset="utf-8">
> > +    <meta name="viewport" content="width=device-width, initial-scale=1">
> > +    <title>Js_of_ocaml • TodoMVC</title>
>
> Let's properly write this instead of snake_case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tastejs/todomvc/pull/1339/files#r33399033>.
>

arthurvr replied:
Oh. Alright. Pretty ugly :smile_cat: Thanks @passy!

arthurvr commented:
Any active community member who's able to review the code? I don't think any of us is (or am I wrong?)

passy commented:
I'm at a read-only level of OCaml understanding. I'll give it a more
thorough look, but it would indeed be great to have someone with actual
experience in this space to review.

On Fri, Jun 26, 2015, 22:49 Arthur Verschaeve <notifications@github.com>
wrote:

> Any active community member who's able to review the code? I don't think
> any of us is (or am I wrong?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tastejs/todomvc/pull/1339#issuecomment-115900122>.
>

slegrand45 commented:
The code has been reviewed by @Drup who is a member of the jsoo project. The code style is now much better.

passy commented:
Oh yes, I prefer the new style a lot over the old `let () ... in`. Well done! :)

arthurvr commented on examples/js_of_ocaml/.gitignore:
```diff
@@ -0,0 +1,2 @@
+_build
+*.byte
```
missing a newline

arthurvr commented on examples/js_of_ocaml/.jscsrc:
```diff
@@ -0,0 +1,69 @@
+	"disallowSpacesInsideObjectBrackets": null,
+	"disallowImplicitTypeConversion": ["string"],
+	"safeContextKeyword": "self"
+}
```
Why is this file here? Just use our code style guidelines and just use the `jscsrc` in the root directory.

slegrand45 replied:
I copied the `jscsrc` from the root directory to be able to modify the `excludeFiles` directive:
```
"excludeFiles": ["**/node_modules/**", "**/bower_components/**", "**/js/todomvc.js"],
```

This is the only method i found to exclude the js file from the `jscsrc` linter without directly putting a directive at the top of the js file. Is there any better way to exclude a file from `jscsrc`? It seems that `jscsrc` doesn't support a feature similar to the `.jshintignore` file.

arthurvr replied:
Just add the ignore in the file in the root?

slegrand45 replied:
If it's possible to propose a modification of the `.jscsrc` file (https://github.com/tastejs/todomvc/blob/master/.jscsrc) then maybe it could be something more generic? Something like:

```
"excludeFiles": ["**/node_modules/**", "**/bower_components/**", "**/auto_gen/**"],
```

So if an implementation uses some automatically generated js files, it can put them in this directory to prevent the `jscsrc` checkings.

arthurvr replied:
Nah. Just add the directory you don't want to check in the jscsrc in the root directory and remove this file :)

arthurvr commented on examples/js_of_ocaml/.jshintignore:
```diff
@@ -0,0 +1 @@
+js/todomvc.ml
```
Isn't there a better way to ignore files then per-example `jshintignore` files? @samccone any ideaz?

samccone replied:
hmm wow i have never seen jshintignore but it actually seems perfect for what we want!

samccone replied:
we just need to have our lint task actually read from this file when linting

arthurvr replied:
:+1:

arthurvr commented on examples/js_of_ocaml/build.sh:
```diff
@@ -0,0 +1,17 @@
+# ' js/todomvc.js
+# sed -i '' '1i\
+# /* jshint ignore:start */
+# ' js/todomvc.js
```
Commented out code doesn't make sense. Let's remove it together with lines 9 and 10.

arthurvr commented on examples/js_of_ocaml/build.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+ocamlbuild -use-ocamlfind \
```
this could use a comment to explain what it does.

arthurvr commented on examples/js_of_ocaml/index.html:
```diff
@@ -0,0 +1,15 @@
+		<script src="js/todomvc.js"></script>
+	</body>
+</html>
+
```
remove this empty line.

arthurvr commented on examples/js_of_ocaml/todomvc.ml:
```diff
@@ -0,0 +1,423 @@
+
```
remove dat line (or is it there for a reason?)

slegrand45 replied:
I think it's more readable with this separation line because it clearly separates the `open` directive from the modules. But i guess it's a matter of taste.

arthurvr commented:
Neat work. Thanks.

arthurvr commented on examples/js_of_ocaml/readme.md:
```diff
@@ -0,0 +1,35 @@
+
+Various code improvements from [Gabriel Radanne](https://github.com/Drup).
+
+Based on [Elm implementation](https://github.com/evancz) by Evan Czaplicki.
```
>  https://github.com/evancz

make this a link on the name itself

arthurvr commented on examples/js_of_ocaml/build.sh:
```diff
@@ -0,0 +1,11 @@
+  todomvc.byte ;
+
+# Build JS code from the OCaml bytecode
+js_of_ocaml +weak.js --opt 3 -o js/todomvc.js todomvc.byte
```
Can you add a description to the readme about how to bulid the app? That makes it clear for people and helps us to maintain the app further on.

arthurvr commented on examples/js_of_ocaml/build.sh:
```diff
@@ -0,0 +1,11 @@
+
+# Compile OCaml source file to OCaml bytecode
+ocamlbuild -use-ocamlfind \
+  -tags "warn(-40)" \
```
Wrong indentation. Check the editorconfig :)

arthurvr commented on examples/js_of_ocaml/readme.md:
```diff
@@ -0,0 +1,35 @@
+If you want to build the application on your own:
+
+1. Install Js_of_ocaml. An easy way is to use [opam](https://opam.ocaml.org/).
+2. See the `build.sh` script.
```
I'd like to extend this description a bit on how exactly to run the `build.sh` script.

arthurvr commented on examples/js_of_ocaml/readme.md:
```diff
@@ -0,0 +1,35 @@
+
+## Credit
+
+Created by [Stéphane Legrand](https://stephanelegrand.wordpress.com/).
```
Try to be consistent. Here you're linking to your personal site while for other people you link to their GitHub profile :)

arthurvr commented on examples/js_of_ocaml/readme.md:
```diff
@@ -0,0 +1,35 @@
+- [GitHub](https://github.com/ocsigen/js_of_ocaml)
+- [Try Js_of_ocaml](http://try.ocamlpro.com/js_of_ocaml/)
+
+### Support
```
`##`

arthurvr commented:
Thanks for your nice work!

samccone commented on examples/js_of_ocaml/index.html:
```diff
@@ -0,0 +1,14 @@
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+	</head>
+	<body>
+		<div id="todomvc"></div>
```
it looks like you are using the 2.0 spec however this root element is an ID, can you switch it to a class?

slegrand45 replied:
Is this an absolute requirement? The standard way with js_of_ocaml is to select first an element by its id and then use it to plug the application. I could select the `<div>` element with the `getElementsByClassName` method but unfortunately this method is for now only available in the development version of js_of_ocaml.

samccone replied:
Is this an absolute requirement? I do not think so, however our test suite does a lookup on this element to determine what version of the spec to use when testing, so clearly this does pose some issues.

What do you think?

slegrand45 replied:
Could we have an `id` attribute and a `class` attribute? ie something like:
```html
<div id="todomvc" class="todoapp"></div>
```

Otherwise, if the idea is to ban any `id` attribute, i have no solution for now until the next release of js_of_ocaml (which will include the `getElementsByClassName` method).

By the way, can you confirm that the needed class name is `todoapp`? Thanks.

samccone commented:
we look for `todoapp` here is the commit that adds the logic.
https://github.com/tastejs/todomvc/commit/45b15b5428bf2bd4313cd7b7aa31ad73c1327be6

For now doing both is an OK solution, thanks for chatting with me, and thanks for the awesome work.

--

ps I tried compiling this app locally by following your instructions and ran into all kinds of errors when installing dependencies, Do you think you might be able to expand the setup instructions for how to compile this app, thanks!

slegrand45 commented:
I added the "todoapp" class. Thank you for your feedback about your attempt to compile the application. I made a test by starting from a clean environment. And indeed it's not as simple as it seems... So i updated the building instructions. I think it's a lot better now.

samccone commented:
looks like we are in need of a rebase :globe_with_meridians:

slegrand45 commented:
After some fixes, there are still 3 failed checks. They are all from the routing part and they all fail because of "stale element reference: element is not attached to the page document". Not sure if i can apply some workarounds?

samccone commented:
@slegrand45 how recently have you rebased off of master? We Very recently merged a fix https://github.com/tastejs/todomvc/pull/1371

arthurvr commented:
Seems like some unrelated commits got into this PR. Can you rebase and get them out?

slegrand45 commented:
It's strange because i already made a rebase... Anyway, i used the following commands:
```sh
git checkout js_of_ocaml
git fetch upstream
git rebase upstream/master
git push -f
```
Strangely enough, i had several conflicts about my own commits during the rebase (?). I used --skip to resolve them. I had also to use "push -f" because git said that my local branch and "origin/js_of_ocaml" have diverged (?). It seems that the PR is now clean again. But i hope i didn't introduce some mess :worried:

Besides, the CI log still shows some "stale element reference" (?).

samccone commented:
yep @slegrand45 looks like you uncovered some more flake :clap:

samccone commented:
hey @slegrand45 would you mind rebasing this off of master and squashing your commits down?

slegrand45 commented:
Seems ok!

samccone commented:
@passy I will let you have final :+1: on this

passy commented:
@slegrand45 I'm seeing some odd rendering errors among other things. Is the latest version on your branch up to date?

![screenshot from 2015-10-09 20-36-54](https://cloud.githubusercontent.com/assets/9906/10403673/5f0117cc-6ec5-11e5-9b45-f61d796a2a49.png)

slegrand45 commented:
@passy Yes it's the latest version. What are the errors you are seeing?

passy commented:
Rendering is broken like on the screenshot and I've only tried a few things that all weren't quite right, focus lost after creating an item, no selection when entering edit mode etc. Do you see those things on your end too?

slegrand45 commented:
Ok, i see what you mean. Indeed, i have the same problems. I'm working on it right now to fix it.

slegrand45 commented:
@passy I think it should be better now.

passy commented:
@slegrand45 Oh yes, way better! :cactus: I'm checking it now. :)

passy commented:
Passing the integration tests with flying colors! Checking with `drool` now.

![screenshot from 2015-10-10 20-40-39](https://cloud.githubusercontent.com/assets/9906/10412940/4bcf5504-6f8f-11e5-817b-040b420e4248.png)

![runningcolourdog](https://cloud.githubusercontent.com/assets/9906/10412947/6b1e13fa-6f8f-11e5-999b-abfb4c1d9e16.gif)

passy commented:
I took me way longer than it should have to find this GIF, but it was worth it and I'm super excited about landing this. Thanks so much @slegrand45! Also thanks @arthurvr and @samccone for the review! :heart_eyes_cat:

![](http://im.ezgif.com/tmp/ezgif-1590176596.gif)

slegrand45 commented:
Thank you very much to all the team! :smiley:

LouisAyroles commented:
Hello everyone, note sure if it's the right place to ask this :/

Do you have an example with version of dependencies ?
Because I have serious problem to make it work..

When I'm trying with latest opam,eliom config I have problem with js_of_ocaml.deriving.syntax not found.

I've tried many different ways and I'm still stucked..

Thx for yout help. :smile:

slegrand45 commented:
Hi,

Indeed, some packages have been renamed since 2015. I will try to send a PR to fix that. Meanwhile, you can take a look at the examples at [https://github.com/slegrand45/examples_ocsigen](https://github.com/slegrand45/examples_ocsigen). These examples should compile and run without any problem. If not, please send an issue report.
